### ISSUE TITLE
Application escape key mode using Ctrl - [ (Ctrl-Bracket)

### ISSUE THREAD

**Issue by GoogleCodeExporter** [mintty/mintty#405]
```
I absolutely love mintty--it's worked wonders for so many years. I've switched 
over to mintty & cygwin vim as my default IDE for all my projects. 

Over the past few months I've got accustomed to escaping insert mode via 
'Ctrl-[' which essentially is the raw ESCAPE code.

Anyway, I was recently reading how to enable Application escape key mode for 
Vim and this works great when using the ESC key, but not with 'Ctrl-['. 

Is there anyway please to support Ctrl-[ sending the application escape code? 
(\e0[)?
```

Original issue reported on code.google.com by `saama...@gmail.com` on 12 Sep 2013 at 6:36

**Comment by mintty** [comment]
Understanding your issue, I consider Ctrl-[ to be somewhat "raw" as opposed to the Esc key so I think in Application Escape mode it would be especially useful to have Ctrl-[ as an alternative to *not* send the application sequence but rather the ESC character.
So unless you would ask for an *additional* mode, I would not like to integrate your proposal.

**Comment by mintty** [comment]
Let me repeat with another example why I think Control-key combinations should not be remapped along with similar remappings for special keys: Alt-Enter toggles full-screen mode, while Alt-Control-M still sends the control characters (ESC ^M) and it's good to have this distinction.
So I'm closing the issue now.
If someone explicit asks for an additional option to also remap Control-key combinations, I may consider to reopen it, but it will never be the default.

**Comment by Karmaflute** [comment]
I also use Ctrl-[ and Ctrl-] religiously in Windows IDEs for code navigation. I would love to be able to bind Ctrl-[ this in Emacs to match my workflow.

**Comment by mintty** [comment]
Is it possible to simply bind the ESC character (Ctrl-[) to a function in Emacs? (Would other escape sequences still be recognised?)
If so, you could also activate application escape key mode and still distinguish the ESC key (then a sequence) to something else.
Or you could use modifyOtherKeys mode (activated with ESC[>4;2m) to turn all control-keys into escape sequences. It would be good, though, for that mode if the application deals with it (like my editor mined :) ).
Or we could talk about a new control setting to remap the ESC character (as opposed to the ESC key for which a setting exists), or maybe any control character if needed.

**Comment by Karmaflute** [comment]
It is not possible to bind the character combination `Ctrl-[` to a function in Emacs unless you're using a GUI version like xemacs or emacs-w32. This is discussed in the following StackExchange threads:

http://emacs.stackexchange.com/questions/7832/how-to-bind-c-for-real
http://emacs.stackexchange.com/questions/10271/how-to-bind-c-for-real-seriously-for-real-this-time

As emacs considers `Ctrl-[` and `Esc` to be the same, this wreaks havok with many things as the `Esc` key is used often. It's also reported that this breaks `M-` as well which would petty much break all of Emacs.

Remapping `Ctrl-[` to an escape sequence seems like the only possibility. `Ctrl-i` would also be nice if you're going to be touching that code anyways.

You mentioned modifyOtherKeys. I don't see an option for this and I'm not sure how I should use your suggestion of `ESC[>4;2m`.

**Comment by mintty** [comment]
This is the plan:

An application control-key mode would be switched by a sequence like 
`\e[?77027h` for Control+[ (ESC character) or `\e[?7709h` for Control+I (TAB character).
These would then send the same key codes as in "modifyOtherKeys" mode (enabled with `\e[>4;2m`),
which would be `\e[91;5u' or `^[[105;5u`, respectively.

Any comments? Is remapping of Control+] needed too?

**Comment by mintty** [comment]
Released in 2.2.1.
